runtime/compiler/control/CompilationThread.cpp

```diff
@@ -2641,35 +2644,89 @@ class ReleaseVMAccessAndAcquireMonitor
    TR::Monitor *_monitor;
    };
 
-void TR::CompilationInfo::prepareForCheckpoint()
+/* IMPORTANT: There should be no return, or C++ exception thrown, after
+ *            releasing the Comp Monitor below until it is re-acquired.
+ *            The Comp Monitor may be acquired in an OMR::CriticalSection
+ *            object; a return, or thrown exception, will destruct this
+ *            object as part leaving the scope, or stack unwinding, which
+ *            will attempt to release the monitor in its destructor.
+ */
+void TR::CompilationInfo::releaseCompMonitorUntilNotifiedOnCRMonitor(J9VMThread *vmThread) throw()
    {
-   J9JavaVM   *vm       = _jitConfig->javaVM;
-   J9VMThread *vmThread = vm->internalVMFunctions->currentVMThread(vm);
+   /* Acquire the CR monitor */
+   acquireCRMonitor();
 
-   if (TR::Options::getCmdLineOptions()->getVerboseOption(TR_VerboseCheckpointRestore))
-      TR_VerboseLog::writeLineLocked(TR_Vlog_CHECKPOINT_RESTORE, "Preparing for checkpoint");
+   /* Release the Comp Monitor before waiting */
+   releaseCompMonitor(vmThread);
```

Comment:
Should we have an assert here that we hold the compMonitor?

Reply:
Updated in https://github.com/eclipse-openj9/openj9/compare/d5850691d5686f23e30b6437904cf504f3a2a87e..fac845ac4f51b4ee18576fb1eb00fb9e002af68d